http-client/Network/HTTP/Client/Connection.hs

```diff
@@ -183,11 +188,28 @@ openSocket tweakSocket addr =
             tweakSocket sock
             NS.connect sock (NS.addrAddress addr)
             return sock)
-
+        
+-- Pick up an IP using an approximation of the happy-eyeballs algorithm:
+-- https://datatracker.ietf.org/doc/html/rfc8305
+-- 
 firstSuccessful :: [NS.AddrInfo] -> (NS.AddrInfo -> IO a) -> IO a
-firstSuccessful []     _  = error "getAddrInfo returned empty list"
-firstSuccessful (a:as) cb =
-    cb a `E.catch` \(e :: E.IOException) ->
-        case as of
-            [] -> E.throwIO e
-            _  -> firstSuccessful as cb
+firstSuccessful []        _  = error "getAddrInfo returned empty list"
+firstSuccessful addresses cb = do             
+    result <- newEmptyMVar
+    either E.throwIO pure =<< 
+        withAsync (tryAddresses result) 
+            (\_ -> takeMVar result)    
+  where
+    -- https://datatracker.ietf.org/doc/html/rfc8305#section-5
+    connectionAttemptDelay = 250 * 1000    
+
+    tryAddresses lastResult = do         
+        z <- forConcurrently (zip addresses [0..]) $ \(addr, n) -> do
+                when (n > 0) $ threadDelay $ n * connectionAttemptDelay        
```

Comment:
Just thinking this through a little bit more, the current behavior is:

* Fork N threads immediately
* N-1 of those threads sleep for 250us

What if instead:

* Fork two threads immediately
* First thread immediately calls `cb addr` on the first address
* Second thread sleeps for 250us and then forks the other N-1 threads

Reply:
Would it make sense to apply this logic recursively?

This version works and I don't see what may be wrong with (except that it can be made better with `foldM` probably).

```
    tryAddresses lastResult = do        
        lastException <- take 1 . reverse . filter isLeft <$> go addresses
        for_ lastException $ tryPutMVar lastResult
      where
        go []             = pure []
        go (addr : addrs) = 
            uncurry (:) <$>
                concurrently 
                    (tryAddress addr) 
                    (do 
                        unless (null addrs) (threadDelay connectionAttemptDelay) 
                        go addrs)

        tryAddress addr = do
            r :: Either E.IOException a <- E.try $ cb addr
            for_ r $ \_ -> tryPutMVar lastResult r
            pure r
 ```

Reply:
Ah no, I see what's wrong with it -- we create extra threads just for `tryAddress addr` which defeats the purpose.

So we can indeed do only one fork, i.e.

```
    tryAddresses result = do
        let addr : addrs = addresses
        z <- concurrently
                (tryAddress addr)
                (do
                    threadDelay connectionAttemptDelay
                    forConcurrently (zip addrs [0..]) $ \(addr, n) -> do
                        when (n > 0) $ threadDelay $ n * connectionAttemptDelay
                        tryAddress addr)
        
        let lastException = take 1 $ reverse $ filter isLeft $ uncurry (:) z
        for_ lastException $ tryPutMVar result
      where
        tryAddress addr = do
            r :: Either E.IOException a <- E.try $! cb addr
            for_ r $ \_ -> tryPutMVar result r
            pure r     
```            

Or create threads only when needed, like this.

```
    tryAddresses result = do
        lastException <- take 1 . filter isLeft <$> go addresses []
        -- If we are here then there were no successful addresses
        for_ lastException $ tryPutMVar result
      where
        go []             asyncs = mapM wait asyncs
        go (addr : addrs) asyncs =
            withAsync (tryAddress addr) $ \a -> do
                unless (null addrs) $ threadDelay connectionAttemptDelay
                go addrs $ a : asyncs

        tryAddress addr = do
            r :: Either E.IOException a <- E.try $ cb addr
            for_ r $ \_ -> tryPutMVar result r
            pure r
```            

I would prefer the latter, since it only create as many threads as actually required.

Reply:
I'm not following this code anymore. It also seems like it's missing some pieces, like the `async` calls.

Reply:
Sorry, I didn't push the code then, was just theorizing on how and when to create threads.
I pushed the commit now, with the latest version of `tryAddresses` using `withAsync`.